src/typing/constraint_js.ml

```diff
@@ -529,21 +529,34 @@ let copy_node node = match node with
     Root { rank; constraints = Unresolved (copy_bounds bounds) }
   | _ -> node
 
+type scope_kind =
+  | VarScope (* var, functions hoisted up to this point *)
+  | LexicalScope (* let, const, classes *)
+
 type scope_entry = {
   specific: Type.t;
   general: Type.t;
   def_loc: Spider_monkey_ast.Loc.t option;
   for_type: bool;
+  constant: bool;
+  scope_kind: scope_kind;
```

Comment:
Let-declarations in a var-scope behave differently than in a let-scope.

```
function x {
  let x = "x";
  let x = "y"; // ok
}
```

```
{
  let x = "x";
  let x = "y"; // error
}
```

That's why I store the scope_kind on both, but I think @jeffmo made a good point above. By using the same data type for both scopes and entries, I'm actually conflating two different concepts. I think it would be more useful to have a `binding_type` as Jeff suggested, which would come with other niceties like removing `for_type` and `constant` optional parameters, and just modeling the domain better.